File: src/leetCodeProblems/BinarySearch/FirstBadVersion278.java
package leetCodeProblems.BinarySearch;

/**
 * LeetCode - https://leetcode.com/problems/first-bad-version
 *
 * TimeComplexity - O(logn)
 * SpaceComplexity - O(1)
 */
public class Version278{

    int badVersion;

    Version278(int badVersion) {
        this.badVersion = badVersion;
    }

    private boolean isBadVersion(int version) {
        if (version == badVersion) {
            return true;
        }

        return false;
    }

    public int Version(int n) {

        int left = 1;
        int right = n;
        int mid = 0;

        while (left < right) {

            mid = left+(right-left)/2;

            //System.out.println("Mid ->" + mid);
            if (isBadVersion(mid)) {
                right = mid;
            }
            else {
                left = mid+1;
            }
        }

        return left;
    }

    public static void main(String[] args) {

        Version278 obj = new Version278(2);

        System.out.println(obj.Version(5));
    }
}

File: src/leetCodeProblems/BinarySearch/FirstBadVersion278.go
package main

import "log"

/*
- LeetCode - https://leetcode.com/problems/first-bad-version/description/
- Time - O(logn)
- Space - O(1)
*/

func isBadVersion(version int) bool {
	return version == 4
}

func firstBadVersion(n int) int {
	start := 0
	end := n

	for start < end {
		mid := start + (end-start)/2

		if isBadVersion(mid) {
			end = mid
		} else {
			start = mid + 1
		}
	}

	return start
}

func main() {
	log.Println(firstBadVersion(5))
}
